File: src/dolphin_context_actions/converters/video.py
import os
import re
import subprocess
import tempfile
import time
from pathlib import Path

from .. import ui
from . import ffmpeg_tools, get_duration, unique_output

GIF_PRESETS = {
    "small": (10, 480),
    "medium": (15, 800),
    "large": (24, 1280),
    "source": (15, -1),
}

GIF_PRESET_NAMES = {
    "small": "Small (10 fps, 480px)",
    "medium": "Medium (15 fps, 800px)",
    "large": "Large (24 fps, 1280px)",
    "source": "Source size (15 fps)",
}


def video_filters(fps: int, width: int) -> str:
    parts = [f"fps={fps}"]
    if width != -1:
        parts.append(f"scale={width}:-1:flags=lanczos")
    return ",".join(parts)


def to_gif(files: list[str], preset: str):
    if preset not in GIF_PRESETS:
        ui.error_dialog("Video → GIF", f"Unknown GIF preset: {preset}")
        return
    fps, width = GIF_PRESETS[preset]

    ffmpeg_bin = ffmpeg_tools.resolve("gif")
    if not ffmpeg_bin:
        ui.error_dialog("Video → GIF", ffmpeg_tools.missing_encoder_message("GIF", "gif"))
        return

    total = len(files)
    errors = []
    done = 0
    handle = ui.pbar_open(
        f"Video → GIF - {GIF_PRESET_NAMES[preset]}",
        f"Starting… (0 of {total})",
    )

    for idx, filepath in enumerate(files):
        input_path = Path(filepath)
        if not input_path.exists():
            errors.append(f"File not found: {filepath}")
            continue

        output_path = input_path.with_suffix(".gif")
        if output_path == input_path:
            output_path = input_path.with_stem(input_path.stem + "_gif").with_suffix(".gif")
        output_path = unique_output(output_path)

        short = input_path.name[:50]
        label = f"[{idx + 1}/{total}] {short}"
        slice_start = int(idx / total * 100)
        slice_end = int((idx + 1) / total * 100)
        palette_end = slice_start + int((slice_end - slice_start) * 0.15)

        ui.pbar_set(handle, slice_start, f"Building palette: {label}")

        duration = get_duration(filepath)
        palette_fd, palette_path = tempfile.mkstemp(prefix="dca_pal_", suffix=".png")
        os.close(palette_fd)

        vf_base = video_filters(fps, width)
        vf_pal = f"{vf_base},palettegen=stats_mode=diff"
        pal_cmd = [ffmpeg_bin, "-y", "-i", filepath, "-vf", vf_pal, "-loglevel", "error", palette_path]
        pal_proc = subprocess.run(pal_cmd, capture_output=True)
        if pal_proc.returncode != 0:
            errors.append(f"{input_path.name} (palette):\n{pal_proc.stderr.decode(errors='replace')[:400]}")
            _cleanup(palette_path)
            continue

        alive = ui.pbar_set(handle, palette_end, f"Converting: {label}")
        if not alive:
            _cleanup(palette_path, output_path)
            ui.pbar_close(handle)
            ui.notify("Video → GIF - Cancelled", f"Cancelled on file {idx + 1}", "dialog-cancel")
            return

        prog_fd, prog_path = tempfile.mkstemp(prefix="dca_vid_", suffix=".txt")
        os.close(prog_fd)
        err_fd, err_path = tempfile.mkstemp(prefix="dca_vid_err_", suffix=".txt")

        lavfi = f"{vf_base}[x];[x][1:v]paletteuse=dither=bayer:bayer_scale=3"
        cmd = [ffmpeg_bin, "-y", "-i", filepath, "-i", palette_path,
               "-lavfi", lavfi, "-progress", prog_path, "-nostats", "-loglevel", "error",
               "--", str(output_path)]

        # stderr to a file, not PIPE: nothing drains a PIPE here, so a full OS
        # pipe buffer would deadlock ffmpeg (the same hang class already hit
        # once with notify-send).
        with os.fdopen(err_fd, "wb") as err_file:
            proc = subprocess.Popen(cmd, stdout=subprocess.DEVNULL, stderr=err_file)
        cancelled = False
        last_pct = palette_end

        while proc.poll() is None:
            time.sleep(0.5)
            # Recomputed every iteration so the cancel check below always
            # runs, even when ffmpeg hasn't written new progress.
            overall = last_pct
            try:
                with open(prog_path) as pf:
                    prog = pf.read()
                matches = re.findall(r"^out_time_ms=(\d+)", prog, re.MULTILINE)
                if matches and duration and duration > 0:
                    us = int(matches[-1])
                    file_pct = min(1.0, us / 1e6 / duration)
                    overall = palette_end + int(file_pct * (slice_end - palette_end))
            except Exception:
                pass
            if overall > last_pct:
                last_pct = overall
            alive = ui.pbar_set(handle, last_pct, f"Converting: {label}")
            if not alive:
                proc.kill()
                cancelled = True
                break

        proc.wait()
        _cleanup(prog_path, palette_path)

        if cancelled:
            _cleanup(output_path, err_path)
            ui.pbar_close(handle)
            ui.notify("Video → GIF - Cancelled", f"Cancelled on file {idx + 1}", "dialog-cancel")
            return

        if proc.returncode != 0:
            try:
                stderr = Path(err_path).read_bytes()
            except Exception:
                stderr = b""
            errors.append(f"{input_path.name}:\n{stderr.decode(errors='replace')[:400]}")
            _cleanup(output_path)
        else:
            done += 1
            ui.pbar_set(handle, slice_end, f"Done: {label}")
        _cleanup(err_path)

    ui.pbar_close(handle)

    if errors:
        preview = "\n\n".join(errors[:3])
        if len(errors) > 3:
            preview += f"\n\n…and {len(errors) - 3} more"
        ui.error_dialog("Video → GIF - Errors", f"Converted {done} of {total}.\n\nErrors:\n{preview}")
        ui.notify("Video → GIF - Finished with errors", f"{done}/{total} converted", "dialog-error")
    elif done > 0:
        ui.notify("Video → GIF - Done", f"✔ {done} file{'s' if done != 1 else ''} → GIF", "image-gif")


def to_mp4(files: list[str], target_ext: str = ".mp4"):
    _transcode_to(files, target_ext, {
        ".mp4": {"codec_v": "libx264", "codec_a": "aac", "opts": ["-preset", "medium", "-crf", "23"]},
        ".webm": {"codec_v": "libvpx-vp9", "codec_a": "libopus", "opts": ["-crf", "30", "-b:v", "0"]},
        ".mkv": {"codec_v": "libx264", "codec_a": "aac", "opts": ["-preset", "medium", "-crf", "23"]},
    }[target_ext], f"→ {target_ext.upper()[1:]}")


def _transcode_to(files: list[str], target_ext: str, params: dict, label_suffix: str):
    ffmpeg_bin = ffmpeg_tools.resolve(params["codec_v"], params["codec_a"])
    if not ffmpeg_bin:
        ui.error_dialog(
            f"Video{label_suffix}",
            ffmpeg_tools.missing_encoder_message(
                target_ext.upper()[1:], params["codec_v"], params["codec_a"]
            ),
        )
        return

    total = len(files)
    errors = []
    done = 0
    handle = ui.pbar_open(f"Video{label_suffix}", f"Starting… (0 of {total})")

    for idx, filepath in enumerate(files):
        input_path = Path(filepath)
        if not input_path.exists():
            errors.append(f"File not found: {filepath}")
            continue

        output_path = input_path.with_suffix(target_ext)
        if output_path == input_path:
            output_path = input_path.with_stem(input_path.stem + "_converted").with_suffix(target_ext)
        output_path = unique_output(output_path)

        short = input_path.name[:50]
        label = f"[{idx + 1}/{total}] {short}"
        slice_start = int(idx / total * 100)
        slice_end = int((idx + 1) / total * 100)

        ui.pbar_set(handle, slice_start, f"Converting: {label}")
        duration = get_duration(filepath)

        prog_fd, prog_path = tempfile.mkstemp(prefix="dca_tr_", suffix=".txt")
        os.close(prog_fd)
        err_fd, err_path = tempfile.mkstemp(prefix="dca_tr_err_", suffix=".txt")

        cmd = [ffmpeg_bin, "-y", "-i", filepath,
               "-c:v", params["codec_v"],
               "-c:a", params["codec_a"],
               *params["opts"],
               "-progress", prog_path, "-nostats", "-loglevel", "error",
               "--", str(output_path)]

        # stderr to a file, not PIPE: nothing drains a PIPE here, so a full OS
        # pipe buffer would deadlock ffmpeg.
        with os.fdopen(err_fd, "wb") as err_file:
            proc = subprocess.Popen(cmd, stdout=subprocess.DEVNULL, stderr=err_file)
        cancelled = False
        last_pct = slice_start

        while proc.poll() is None:
            time.sleep(0.5)
            # Recomputed every iteration so the cancel check below always
            # runs, even when ffmpeg hasn't written new progress.
            overall = last_pct
            try:
                with open(prog_path) as pf:
                    prog = pf.read()
                matches = re.findall(r"^out_time_ms=(\d+)", prog, re.MULTILINE)
                if matches and duration and duration > 0:
                    us = int(matches[-1])
                    pct = min(1.0, us / 1e6 / duration)
                    overall = slice_start + int(pct * (slice_end - slice_start))
            except Exception:
                pass
            if overall > last_pct:
                last_pct = overall
            alive = ui.pbar_set(handle, last_pct, f"Converting: {label}")
            if not alive:
                proc.kill()
                cancelled = True
                break

        proc.wait()
        _cleanup(prog_path)

        if cancelled:
            _cleanup(output_path, err_path)
            ui.pbar_close(handle)
            ui.notify(f"Video{label_suffix} - Cancelled", f"Cancelled on file {idx + 1}", "dialog-cancel")
            return

        if proc.returncode != 0:
            try:
                stderr = Path(err_path).read_bytes()
            except Exception:
                stderr = b""
            errors.append(f"{input_path.name}:\n{stderr.decode(errors='replace')[:400]}")
            _cleanup(output_path)
        else:
            done += 1
            ui.pbar_set(handle, slice_end, f"Done: {label}")
        _cleanup(err_path)

    ui.pbar_close(handle)

    if errors:
        preview = "\n\n".join(errors[:3])
        if len(errors) > 3:
            preview += f"\n\n…and {len(errors) - 3} more"
        ui.error_dialog(f"Video{label_suffix} - Errors", f"Converted {done} of {total}.\n\nErrors:\n{preview}")
    elif done > 0:
        ui.notify(f"Video{label_suffix} - Done", f"✔ {done} file{'s' if done != 1 else ''}{label_suffix}", "video-x-generic")


def extract_audio(files: list[str]):
    ffmpeg_bin = ffmpeg_tools.resolve("libmp3lame")
    if not ffmpeg_bin:
        ui.error_dialog("Extract Audio", ffmpeg_tools.missing_encoder_message("MP3", "libmp3lame"))
        return

    total = len(files)
    errors = []
    done = 0
    handle = ui.pbar_open("Extract Audio", f"Starting… (0 of {total})")

    for idx, filepath in enumerate(files):
        input_path = Path(filepath)
        if not input_path.exists():
            errors.append(f"File not found: {filepath}")
            continue

        output_path = input_path.with_suffix(".mp3")
        if output_path == input_path:
            output_path = input_path.with_stem(input_path.stem + "_audio").with_suffix(".mp3")
        output_path = unique_output(output_path)

        short = input_path.name[:50]
        label = f"[{idx + 1}/{total}] {short}"
        slice_start = int(idx / total * 100)
        slice_end = int((idx + 1) / total * 100)

        ui.pbar_set(handle, slice_start, f"Extracting: {label}")
        duration = get_duration(filepath)

        prog_fd, prog_path = tempfile.mkstemp(prefix="dca_ext_", suffix=".txt")
        os.close(prog_fd)
        err_fd, err_path = tempfile.mkstemp(prefix="dca_ext_err_", suffix=".txt")

        cmd = [ffmpeg_bin, "-y", "-i", filepath,
               "-vn", "-c:a", "libmp3lame", "-aq", "2",
               "-progress", prog_path, "-nostats", "-loglevel", "error",
               "--", str(output_path)]

        # stderr to a file, not PIPE: nothing drains a PIPE here, so a full OS
        # pipe buffer would deadlock ffmpeg.
        with os.fdopen(err_fd, "wb") as err_file:
            proc = subprocess.Popen(cmd, stdout=subprocess.DEVNULL, stderr=err_file)
        cancelled = False
        last_pct = slice_start

        while proc.poll() is None:
            time.sleep(0.5)
            # Recomputed every iteration so the cancel check below always
            # runs, even when ffmpeg hasn't written new progress.
            overall = last_pct
            try:
                with open(prog_path) as pf:
                    prog = pf.read()
                matches = re.findall(r"^out_time_ms=(\d+)", prog, re.MULTILINE)
                if matches and duration and duration > 0:
                    us = int(matches[-1])
                    pct = min(1.0, us / 1e6 / duration)
                    overall = slice_start + int(pct * (slice_end - slice_start))
            except Exception:
                pass
            if overall > last_pct:
                last_pct = overall
            alive = ui.pbar_set(handle, last_pct, f"Extracting: {label}")
            if not alive:
                proc.kill()
                cancelled = True
                break

        proc.wait()
        _cleanup(prog_path)

        if cancelled:
            _cleanup(output_path, err_path)
            ui.pbar_close(handle)
            ui.notify("Extract Audio - Cancelled", f"Cancelled on file {idx + 1}", "dialog-cancel")
            return

        if proc.returncode != 0:
            try:
                stderr = Path(err_path).read_bytes()
            except Exception:
                stderr = b""
            errors.append(f"{input_path.name}:\n{stderr.decode(errors='replace')[:400]}")
            _cleanup(output_path)
        else:
            done += 1
            ui.pbar_set(handle, slice_end, f"Done: {label}")
        _cleanup(err_path)

    ui.pbar_close(handle)

    if errors:
        preview = "\n\n".join(errors[:3])
        if len(errors) > 3:
            preview += f"\n\n…and {len(errors) - 3} more"
        ui.error_dialog("Extract Audio - Errors", f"Extracted {done} of {total}.\n\nErrors:\n{preview}")
    elif done > 0:
        ui.notify("Extract Audio - Done", f"✔ {done} file{'s' if done != 1 else ''} → MP3", "audio-x-mp3")


def _cleanup(*paths):
    for p in paths:
        try:
            Path(p).unlink()
        except Exception:
            pass

File: src/converters/video.rs
use std::fs;
use std::path::Path;
use std::process::{Command, Stdio};

use tempfile::NamedTempFile;

use super::{
    ffmpeg_tools, get_duration, run_ffmpeg_progress, unique_output, FfmpegOutcome,
};
use crate::ui;

pub const GIF_PRESET_NAMES: &[(&str, &str)] = &[
    ("small", "Small (10 fps, 480px)"),
    ("medium", "Medium (15 fps, 800px)"),
    ("large", "Large (24 fps, 1280px)"),
    ("source", "Source size (15 fps)"),
];

fn gif_preset(name: &str) -> Option<(i32, i32)> {
    match name {
        "small" => Some((10, 480)),
        "medium" => Some((15, 800)),
        "large" => Some((24, 1280)),
        "source" => Some((15, -1)),
        _ => None,
    }
}

pub fn gif_preset_name(key: &str) -> Option<&'static str> {
    GIF_PRESET_NAMES.iter().find(|(k, _)| *k == key).map(|(_, n)| *n)
}

fn video_filters(fps: i32, width: i32) -> String {
    if width == -1 {
        format!("fps={fps}")
    } else {
        format!("fps={fps},scale={width}:-1:flags=lanczos")
    }
}

fn cleanup(paths: &[&Path]) {
    for path in paths {
        let _ = fs::remove_file(path);
    }
}

pub fn to_gif(files: &[String], preset: &str) {
    let Some((fps, width)) = gif_preset(preset) else {
        ui::error_dialog("Video → GIF", &format!("Unknown GIF preset: {preset}"));
        return;
    };
    let Some(ffmpeg_bin) = ffmpeg_tools::resolve(&["gif"]) else {
        ui::error_dialog("Video → GIF", &ffmpeg_tools::missing_encoder_message("GIF", &["gif"]));
        return;
    };

    let total = files.len();
    let mut errors = Vec::new();
    let mut done = 0usize;
    let title = format!("Video → GIF - {}", gif_preset_name(preset).unwrap_or(preset));
    let handle = ui::pbar_open(&title, &format!("Starting… (0 of {total})"));

    for (idx, filepath) in files.iter().enumerate() {
        let input_path = Path::new(filepath);
        if !input_path.exists() {
            errors.push(format!("File not found: {filepath}"));
            continue;
        }

        let mut output_path = input_path.with_extension("gif");
        if output_path == input_path {
            let stem = input_path.file_stem().and_then(|s| s.to_str()).unwrap_or("out");
            output_path = input_path.with_file_name(format!("{stem}_gif.gif"));
        }
        output_path = unique_output(&output_path);

        let short = input_path
            .file_name()
            .map(|n| n.to_string_lossy().chars().take(50).collect::<String>())
            .unwrap_or_default();
        let label = format!("[{}/{total}] {short}", idx + 1);
        let slice_start = ((idx as f64 / total as f64) * 100.0) as i32;
        let slice_end = (((idx + 1) as f64 / total as f64) * 100.0) as i32;
        let palette_end = slice_start + ((slice_end - slice_start) as f64 * 0.15) as i32;
        ui::pbar_set(handle.as_ref(), slice_start, Some(&format!("Building palette: {label}")));

        let duration = get_duration(filepath);
        let palette = NamedTempFile::with_suffix(".png").ok();
        let Some(palette) = palette else {
            errors.push(format!("{short}: cannot create palette temp file"));
            continue;
        };
        let palette_path = palette.path().to_path_buf();

        let vf_base = video_filters(fps, width);
        let vf_pal = format!("{vf_base},palettegen=stats_mode=diff");
        let pal_status = Command::new(&ffmpeg_bin)
            .args(["-y", "-i", filepath, "-vf", &vf_pal, "-loglevel", "error"])
            .arg(&palette_path)
            .output();
        match pal_status {
            Ok(out) if out.status.success() => {}
            Ok(out) => {
                let err = String::from_utf8_lossy(&out.stderr);
                errors.push(format!(
                    "{short} (palette):\n{}",
                    err.chars().take(400).collect::<String>()
                ));
                continue;
            }
            Err(e) => {
                errors.push(format!("{short} (palette):\n{e}"));
                continue;
            }
        }

        if !ui::pbar_set(handle.as_ref(), palette_end, Some(&format!("Converting: {label}"))) {
            cleanup(&[&output_path]);
            ui::pbar_close(handle.as_ref());
            ui::notify(
                "Video → GIF - Cancelled",
                &format!("Cancelled on file {}", idx + 1),
                "dialog-cancel",
            );
            return;
        }

        let prog = NamedTempFile::new().ok();
        let mut err = NamedTempFile::new().ok();
        let prog_path = prog.as_ref().map(|f| f.path().to_path_buf());
        let lavfi = format!("{vf_base}[x];[x][1:v]paletteuse=dither=bayer:bayer_scale=3");
        let mut cmd = Command::new(&ffmpeg_bin);
        cmd.args(["-y", "-i", filepath, "-i"]);
        cmd.arg(&palette_path);
        cmd.args(["-lavfi", &lavfi]);
        if let Some(prog_path) = &prog_path {
            cmd.args(["-progress", &prog_path.display().to_string()]);
        }
        cmd.args(["-nostats", "-loglevel", "error", "--"]);
        cmd.arg(&output_path);
        cmd.stdout(Stdio::null());
        if let Some(err_file) = err.as_mut() {
            if let Ok(clone) = err_file.as_file().try_clone() {
                cmd.stderr(clone);
            }
        }

        let proc = match cmd.spawn() {
            Ok(p) => p,
            Err(e) => {
                errors.push(format!("{short}:\n{e}"));
                continue;
            }
        };
        let outcome = run_ffmpeg_progress(
            proc,
            prog_path.as_deref(),
            duration,
            handle.as_ref(),
            palette_end,
            slice_end,
            &format!("Converting: {label}"),
        );
        drop(prog);
        drop(palette);

        match outcome {
            FfmpegOutcome::Cancelled => {
                cleanup(&[&output_path]);
                ui::pbar_close(handle.as_ref());
                ui::notify(
                    "Video → GIF - Cancelled",
                    &format!("Cancelled on file {}", idx + 1),
                    "dialog-cancel",
                );
                return;
            }
            FfmpegOutcome::Failed(_) => {
                let stderr = err
                    .as_ref()
                    .and_then(|f| fs::read(f.path()).ok())
                    .unwrap_or_default();
                let text = String::from_utf8_lossy(&stderr);
                errors.push(format!("{short}:\n{}", text.chars().take(400).collect::<String>()));
                cleanup(&[&output_path]);
            }
            FfmpegOutcome::Success => {
                done += 1;
                ui::pbar_set(handle.as_ref(), slice_end, Some(&format!("Done: {label}")));
            }
        }
    }

    ui::pbar_close(handle.as_ref());
    finish_batch("Video → GIF", "image-gif", "GIF", done, total, &errors);
}

pub fn to_mp4(files: &[String], target_ext: &str) {
    let params = match target_ext {
        ".mp4" => ("libx264", "aac", &["-preset", "medium", "-crf", "23"][..]),
        ".webm" => ("libvpx-vp9", "libopus", &["-crf", "30", "-b:v", "0"][..]),
        ".mkv" => ("libx264", "aac", &["-preset", "medium", "-crf", "23"][..]),
        _ => {
            ui::error_dialog("Video", &format!("Unknown target: {target_ext}"));
            return;
        }
    };
    let label_suffix = format!("→ {}", target_ext.trim_start_matches('.').to_uppercase());
    transcode_to(files, target_ext, params.0, params.1, params.2, &label_suffix);
}

fn transcode_to(
    files: &[String],
    target_ext: &str,
    codec_v: &str,
    codec_a: &str,
    opts: &[&str],
    label_suffix: &str,
) {
    let Some(ffmpeg_bin) = ffmpeg_tools::resolve(&[codec_v, codec_a]) else {
        ui::error_dialog(
            &format!("Video{label_suffix}"),
            &ffmpeg_tools::missing_encoder_message(
                &target_ext.trim_start_matches('.').to_uppercase(),
                &[codec_v, codec_a],
            ),
        );
        return;
    };

    let total = files.len();
    let mut errors = Vec::new();
    let mut done = 0usize;
    let handle = ui::pbar_open(&format!("Video{label_suffix}"), &format!("Starting… (0 of {total})"));

    for (idx, filepath) in files.iter().enumerate() {
        let input_path = Path::new(filepath);
        if !input_path.exists() {
            errors.push(format!("File not found: {filepath}"));
            continue;
        }
        let mut output_path = input_path.with_extension(target_ext.trim_start_matches('.'));
        if output_path == input_path {
            let stem = input_path.file_stem().and_then(|s| s.to_str()).unwrap_or("out");
            output_path = input_path.with_file_name(format!(
                "{stem}_converted.{}",
                target_ext.trim_start_matches('.')
            ));
        }
        output_path = unique_output(&output_path);

        let short = input_path
            .file_name()
            .map(|n| n.to_string_lossy().chars().take(50).collect::<String>())
            .unwrap_or_default();
        let label = format!("[{}/{total}] {short}", idx + 1);
        let slice_start = ((idx as f64 / total as f64) * 100.0) as i32;
        let slice_end = (((idx + 1) as f64 / total as f64) * 100.0) as i32;
        ui::pbar_set(handle.as_ref(), slice_start, Some(&format!("Converting: {label}")));
        let duration = get_duration(filepath);

        let prog = NamedTempFile::new().ok();
        let mut err = NamedTempFile::new().ok();
        let prog_path = prog.as_ref().map(|f| f.path().to_path_buf());
        let mut cmd = Command::new(&ffmpeg_bin);
        cmd.args(["-y", "-i", filepath, "-c:v", codec_v, "-c:a", codec_a]);
        cmd.args(opts);
        if let Some(prog_path) = &prog_path {
            cmd.args(["-progress", &prog_path.display().to_string()]);
        }
        cmd.args(["-nostats", "-loglevel", "error", "--"]);
        cmd.arg(&output_path);
        cmd.stdout(Stdio::null());
        if let Some(err_file) = err.as_mut() {
            if let Ok(clone) = err_file.as_file().try_clone() {
                cmd.stderr(clone);
            }
        }
        let proc = match cmd.spawn() {
            Ok(p) => p,
            Err(e) => {
                errors.push(format!("{short}:\n{e}"));
                continue;
            }
        };
        let outcome = run_ffmpeg_progress(
            proc,
            prog_path.as_deref(),
            duration,
            handle.as_ref(),
            slice_start,
            slice_end,
            &format!("Converting: {label}"),
        );
        drop(prog);
        match outcome {
            FfmpegOutcome::Cancelled => {
                cleanup(&[&output_path]);
                ui::pbar_close(handle.as_ref());
                ui::notify(
                    &format!("Video{label_suffix} - Cancelled"),
                    &format!("Cancelled on file {}", idx + 1),
                    "dialog-cancel",
                );
                return;
            }
            FfmpegOutcome::Failed(_) => {
                let stderr = err
                    .as_ref()
                    .and_then(|f| fs::read(f.path()).ok())
                    .unwrap_or_default();
                let text = String::from_utf8_lossy(&stderr);
                errors.push(format!("{short}:\n{}", text.chars().take(400).collect::<String>()));
                cleanup(&[&output_path]);
            }
            FfmpegOutcome::Success => {
                done += 1;
                ui::pbar_set(handle.as_ref(), slice_end, Some(&format!("Done: {label}")));
            }
        }
    }

    ui::pbar_close(handle.as_ref());
    if !errors.is_empty() {
        let mut preview = errors.iter().take(3).cloned().collect::<Vec<_>>().join("\n\n");
        if errors.len() > 3 {
            preview.push_str(&format!("\n\n…and {} more", errors.len() - 3));
        }
        ui::error_dialog(
            &format!("Video{label_suffix} - Errors"),
            &format!("Converted {done} of {total}.\n\nErrors:\n{preview}"),
        );
    } else if done > 0 {
        let plural = if done != 1 { "s" } else { "" };
        ui::notify(
            &format!("Video{label_suffix} - Done"),
            &format!("✔ {done} file{plural}{label_suffix}"),
            "video-x-generic",
        );
    }
}

pub fn extract_audio(files: &[String]) {
    let Some(ffmpeg_bin) = ffmpeg_tools::resolve(&["libmp3lame"]) else {
        ui::error_dialog(
            "Extract Audio",
            &ffmpeg_tools::missing_encoder_message("MP3", &["libmp3lame"]),
        );
        return;
    };

    let total = files.len();
    let mut errors = Vec::new();
    let mut done = 0usize;
    let handle = ui::pbar_open("Extract Audio", &format!("Starting… (0 of {total})"));

    for (idx, filepath) in files.iter().enumerate() {
        let input_path = Path::new(filepath);
        if !input_path.exists() {
            errors.push(format!("File not found: {filepath}"));
            continue;
        }
        let mut output_path = input_path.with_extension("mp3");
        if output_path == input_path {
            let stem = input_path.file_stem().and_then(|s| s.to_str()).unwrap_or("out");
            output_path = input_path.with_file_name(format!("{stem}_audio.mp3"));
        }
        output_path = unique_output(&output_path);

        let short = input_path
            .file_name()
            .map(|n| n.to_string_lossy().chars().take(50).collect::<String>())
            .unwrap_or_default();
        let label = format!("[{}/{total}] {short}", idx + 1);
        let slice_start = ((idx as f64 / total as f64) * 100.0) as i32;
        let slice_end = (((idx + 1) as f64 / total as f64) * 100.0) as i32;
        ui::pbar_set(handle.as_ref(), slice_start, Some(&format!("Extracting: {label}")));
        let duration = get_duration(filepath);

        let prog = NamedTempFile::new().ok();
        let mut err = NamedTempFile::new().ok();
        let prog_path = prog.as_ref().map(|f| f.path().to_path_buf());
        let mut cmd = Command::new(&ffmpeg_bin);
        cmd.args([
            "-y",
            "-i",
            filepath,
            "-vn",
            "-c:a",
            "libmp3lame",
            "-aq",
            "2",
        ]);
        if let Some(prog_path) = &prog_path {
            cmd.args(["-progress", &prog_path.display().to_string()]);
        }
        cmd.args(["-nostats", "-loglevel", "error", "--"]);
        cmd.arg(&output_path);
        cmd.stdout(Stdio::null());
        if let Some(err_file) = err.as_mut() {
            if let Ok(clone) = err_file.as_file().try_clone() {
                cmd.stderr(clone);
            }
        }
        let proc = match cmd.spawn() {
            Ok(p) => p,
            Err(e) => {
                errors.push(format!("{short}:\n{e}"));
                continue;
            }
        };
        let outcome = run_ffmpeg_progress(
            proc,
            prog_path.as_deref(),
            duration,
            handle.as_ref(),
            slice_start,
            slice_end,
            &format!("Extracting: {label}"),
        );
        drop(prog);
        match outcome {
            FfmpegOutcome::Cancelled => {
                cleanup(&[&output_path]);
                ui::pbar_close(handle.as_ref());
                ui::notify(
                    "Extract Audio - Cancelled",
                    &format!("Cancelled on file {}", idx + 1),
                    "dialog-cancel",
                );
                return;
            }
            FfmpegOutcome::Failed(_) => {
                let stderr = err
                    .as_ref()
                    .and_then(|f| fs::read(f.path()).ok())
                    .unwrap_or_default();
                let text = String::from_utf8_lossy(&stderr);
                errors.push(format!("{short}:\n{}", text.chars().take(400).collect::<String>()));
                cleanup(&[&output_path]);
            }
            FfmpegOutcome::Success => {
                done += 1;
                ui::pbar_set(handle.as_ref(), slice_end, Some(&format!("Done: {label}")));
            }
        }
    }

    ui::pbar_close(handle.as_ref());
    if !errors.is_empty() {
        let mut preview = errors.iter().take(3).cloned().collect::<Vec<_>>().join("\n\n");
        if errors.len() > 3 {
            preview.push_str(&format!("\n\n…and {} more", errors.len() - 3));
        }
        ui::error_dialog(
            "Extract Audio - Errors",
            &format!("Extracted {done} of {total}.\n\nErrors:\n{preview}"),
        );
    } else if done > 0 {
        let plural = if done != 1 { "s" } else { "" };
        ui::notify(
            "Extract Audio - Done",
            &format!("✔ {done} file{plural} → MP3"),
            "audio-x-mp3",
        );
    }
}

fn finish_batch(
    title: &str,
    icon: &str,
    target: &str,
    done: usize,
    total: usize,
    errors: &[String],
) {
    if !errors.is_empty() {
        let mut preview = errors.iter().take(3).cloned().collect::<Vec<_>>().join("\n\n");
        if errors.len() > 3 {
            preview.push_str(&format!("\n\n…and {} more", errors.len() - 3));
        }
        ui::error_dialog(
            &format!("{title} - Errors"),
            &format!("Converted {done} of {total}.\n\nErrors:\n{preview}"),
        );
        ui::notify(
            &format!("{title} - Finished with errors"),
            &format!("{done}/{total} converted"),
            "dialog-error",
        );
    } else if done > 0 {
        let plural = if done != 1 { "s" } else { "" };
        ui::notify(
            &format!("{title} - Done"),
            &format!("✔ {done} file{plural} → {target}"),
            icon,
        );
    }
}
